src/dkg.rs

```diff
@@ -0,0 +1,413 @@
+//! parties.
+//!
+//! [Pedersen's DKG]: https://link.springer.com/content/pdf/10.1007/3-540-46416-6_47.pdf
+//! [Gennaro et al.]: https://link.springer.com/content/pdf/10.1007/3-540-48910-X_21.pdf
+
+use rand_core::{CryptoRng, RngCore};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use core::fmt;
+use sha2::{Digest, Sha256};
+
+use crate::{
+    alloc::{vec, Vec},
+    group::Group,
+    proofs::ProofOfPossession,
+    sharing::Dealer,
+    sharing::{self, Params, PublicKeySet, PublicPolynomial},
+    PublicKey, SecretKey,
+};
+
+/// Errors that can occur during the secret sharing protocol.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Secret received from the party does not correspond to their commitment via
+    /// the public polynomial.
+    InvalidSecret,
+    /// Provided commitment does not correspond to the party's public key share.
+    InvalidCommitment,
+    /// Secret share for this participant was already provided.
+    DuplicitShare,
+    /// Secret shares from some parties are missing.
+    MissingShares,
+    /// Provided proof of possession or public polynomial is malformed.
+    MalformedParticipantProof(sharing::Error),
+    /// Public shares obtained from cumulated public polynomial are inconsistent
+    InconsistentPublicShares(sharing::Error),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidSecret => formatter.write_str(
+                "secret received from the party does not correspond to their commitment via \
+                public polynomial",
+            ),
+            Self::InvalidCommitment => formatter.write_str(
+                "public polynomial received from one of the parties does not correspond \
+                to their commitment",
+            ),
+            Self::DuplicitShare => {
+                formatter.write_str("secret share for this participant was already provided")
+            }
+            Self::MissingShares => {
+                formatter.write_str("secret shares from some parties are missing")
+            }
+            Self::MalformedParticipantProof(err) => {
+                write!(
+                    formatter,
+                    "provided proof of possession or public polynomial is malformed: {err}"
+                )
+            }
+            Self::InconsistentPublicShares(err) => {
+                write!(
+                    formatter,
+                    "public shares obtained from cumulated public polynomial are inconsistent: {err}"
+                )
+            }
+        }
+    }
+}
+
+/// Structure for collecting commitments of public share in committed Pedersen's
+/// distributed key generation.
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(bound = ""))]
+pub struct DkgCommitmentsCollector<G: Group> {
```

Comment:
Bikeshedding: I'd name types corresponding to the different steps of the DKG protocol something like `ParticipantCollectingCommitments`, `ParticipantCollectingPolynomials` and `ParticipantExchangingSecrets` (provided that the protocol is further split as I've described in another comment). This would reflect the facts that these data types correspond to the protocol participants and contain secret information specific to a participant.